cms/grading/scorecache.py

```diff
@@ -0,0 +1,593 @@
+#!/usr/bin/env python3
+
+# Contest Management System - http://cms-dev.github.io/
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Score cache utilities for AWS ranking performance.
+
+This module provides functions to update and query the score cache,
+which stores pre-computed task scores for each participation to speed
+up ranking page loading.
+
+"""
+
+from datetime import datetime
+
+from sqlalchemy.orm import Session
+
+from cms.db import (
+    Participation, Task, Submission,
+    ParticipationTaskScore, ScoreHistory
+)
+from cmscommon.constants import (
+    SCORE_MODE_MAX, SCORE_MODE_MAX_SUBTASK, SCORE_MODE_MAX_TOKENED_LAST
+)
+
+
+__all__ = [
+    "update_score_cache",
+    "invalidate_score_cache",
+    "rebuild_score_cache",
+    "rebuild_score_history",
+    "get_cached_score_entry",
+]
+
+
+def _parse_subtask_scores(score_details, score: float) -> dict[str, float] | None:
+    """Parse subtask scores from score_details.
+
+    Returns a dict mapping subtask index (as string) to score,
+    or None if score_details indicates no score (empty list with score 0).
+    """
+    if score_details == [] and score == 0.0:
+        return None
+
+    try:
+        subtask_scores = dict(
+            (str(subtask["idx"]), subtask["score"])
+            for subtask in score_details
+        )
+    except Exception:
+        subtask_scores = None
+
+    if subtask_scores is None or len(subtask_scores) == 0:
+        subtask_scores = {"1": score}
+
+    return subtask_scores
+
+
+def _compute_final_score(
```

Comment:
This seems like a close copy of https://github.com/ioi-isr/cms/blob/4e6caa4da7833dbf73ec692f076b9c7230e3dc6e/cms/grading/scoring.py#L106 -- is it possible to reuse it? If not, could it be referred to here?